lib/segment/src/vector_storage/query/context_query.rs

```diff
@@ -0,0 +1,123 @@
+use std::cmp::min;
+
+use common::math::scaled_fast_sigmoid;
+use common::types::ScoreType;
+use ordered_float::OrderedFloat;
+
+use super::discovery_query::DiscoveryPair;
+use super::{Query, TransformInto};
+use crate::data_types::vectors::{QueryVector, VectorType};
+
+impl<T> DiscoveryPair<T> {
+    /// In the first stage of discovery search, the objective is to get the best entry point
+    /// for the search. This is done by using a smooth loss function instead of hard ranking
+    /// to approach the best zone, once the best zone is reached, score will be same for all
+    /// points inside that zone.
+    ///
+    ///                   │
+    ///                   │         
+    ///                   │    +0
+    ///                   │             +0
+    ///                   │
+    ///         n         │         p
+    ///                   │
+    ///   ─►          ─►  │
+    ///  -0.4        -0.1 │   +0
+    ///                   │
+    ///
+    /// Output will always be in the range (-1, 0]
+    pub fn loss_by(&self, similarity: impl Fn(&T) -> ScoreType) -> ScoreType {
+        const MARGIN: ScoreType = ScoreType::EPSILON;
+
+        let positive = scaled_fast_sigmoid(similarity(&self.positive));
+        let negative = scaled_fast_sigmoid(similarity(&self.negative));
+
+        let difference = OrderedFloat(positive - negative - MARGIN);
+
+        *min(difference, OrderedFloat(0.0))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ContextQuery<T> {
+    pub pairs: Vec<DiscoveryPair<T>>,
+}
+
+impl<T> ContextQuery<T> {
+    pub fn new(pairs: Vec<DiscoveryPair<T>>) -> Self {
+        Self { pairs }
+    }
+
+    pub fn iter_all(&self) -> impl Iterator<Item = &T> {
```

Comment:
I find `iter_all` a bit confusing, as I'd expect this to iterate over the pairs by itself. Would something like `iter_elements` or `iter_values` be better?

Maybe there's a good reason to name it this way, then I might have missed it.

Reply:
Thanks for the suggestion, `iter_all` was just the first name that popped into my mind and didn't think it would be confusing. I like `iter_values`